tests/conftest.py

```diff
@@ -26,17 +32,114 @@ def app_params(request):
             kwargs.update(info.kwargs)
 
     args = [pargs[i] for i in sorted(pargs.keys())]
-    return args, kwargs
+
+    # ##### process pytest.mark.test_params
+
+    if test_params['specific_srcdir'] and 'srcdir' not in kwargs:
+        kwargs['srcdir'] = test_params['specific_srcdir']
+
+    if test_params['shared_result']:
+        restore = shared_result.restore(test_params['shared_result'])
+        kwargs.update(restore)
+
+    # ##### prepare Application params
+
+    if 'srcdir' in kwargs:
+        srcdir = util.tempdir / kwargs['srcdir']
+    else:
+        srcdir = util.tempdir / kwargs.get('testroot', 'root')
+    kwargs['srcdir'] = srcdir
+
+    if kwargs.get('testroot') is None:
+        testroot_path = util.rootdir / 'root'
+    else:
+        testroot_path = util.rootdir / 'roots' / ('test-' + kwargs['testroot'])
+
+    if not srcdir.exists():
+        testroot_path.copytree(srcdir)
+
+    return namedtuple('app_params', 'args,kwargs')(args, kwargs)
+
+
+@pytest.fixture
+def test_params(request):
+    """
+    test parameters that is specified by 'pytest.mark.test_params'
+
+    :param Union[str, bool, None] specific_srcdir:
+       If True, testroot directory will be copied into
+       '<TMPDIR>/<TEST FUNCTION NAME>'.
+       If string is specified, it copied into '<TMPDIR>/<THE STRING>'.
+       You can used this feature for providing special crafted source
+       directory. Also you can used for sharing source directory for
+       parametrized testing and/or inter test functions. Default is None.
+    :param Union[str, bool, None] shared_result:
+       If True, app._status and app._warning objects will be shared in the
+       parametrized test functions. If string is specified, the objects will
+       be shred in the test functions that have same 'shared_result' value.
+       If you don't specify specific_srcdir, this option override
+       specific_srcdir param by 'shared_result' value. Default is None.
+    """
+    env = request.node.get_marker('test_params')
+    kwargs = env.kwargs if env else {}
+    result = {
+        'specific_srcdir': None,
+        'shared_result': None,
+    }
+    result.update(kwargs)
+
+    if (result['shared_result'] and
+            not isinstance(result['shared_result'], string_types)):
+        result['shared_result'] = request.node.originalname or request.node.name
+
+    if result['shared_result'] and not result['specific_srcdir']:
+        result['specific_srcdir'] = result['shared_result']
+
+    if (result['specific_srcdir'] and
+        not isinstance(result['specific_srcdir'], string_types)):
+        result['specific_srcdir'] = request.node.originalname or request.node.name
+
+    return result
+
+
+class AppWrapper(object):
```

Comment:
Agreed. Actually I don't like this name very much, but I could not think of a good name too.
This class is not a open API, so long name is also OK.
How about `SphinxTestAppWrapperForSkipBuilding` ?